adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+
+    Args:
+        model: model name. Refer to litellm supported models for how to pass
+               this: https://litellm.vercel.app/docs/providers
+        user_prompt: User prompt.
+        system_prompt: System prompt.
+        messages: List of messages to be sent to the model. If provided, `user_prompt`, `system_prompt` and `instruction_first` will be ignored.
+        api_key: API key, optional. If provided, will be used to authenticate
+                 with the provider of your specified model.
+        instruction_first: Whether to put instructions first.
+        response_model: Pydantic model to constrain the LLM generated response. If not provided, the raw completion text will be returned.  # noqa
+        max_tokens: Maximum tokens to generate.
+        temperature: Temperature for sampling.
+        timeout: Timeout in seconds.
+
+    Returns:
+        Dict[str, Any]: OpenAI response or error message.
+    """
+
+    if not user_prompt and not messages:
+        raise ValueError("You must provide either `user_prompt` or `messages`.")
+
+    if not messages:
+        # get messages from user_prompt and system_prompt
+        messages = get_messages(user_prompt, system_prompt, instruction_first)
+    response = LLMResponse()
+
+    if response_model is None:
+        # unconstrained generation - return raw completion text and store it in `data` field: {"text": completion_text}
+        try:
+            completion = await litellm.acompletion(
+                model=model,
+                api_key=api_key,
+                messages=messages,
+                max_tokens=max_tokens,
+                temperature=temperature,
+                timeout=timeout,
+            )
+            completion_text = completion.choices[0].message.content
+            response.data['_completion_text'] = completion_text
+        except Exception as e:
+            response.adala_error = True
+            response.adala_message = type(e).__name__
+            # create a traceback string
+            response.adala_details = traceback.format_exc()
+        finally:
+            return response.model_dump(by_alias=True)
+
+    # constrained generation branch - use `response_model` to constrain the LLM response
+    try:
+        instructor_response, completion = await async_instructor_client.chat.completions.create_with_completion(
+            model=model,
+            api_key=api_key,
+            messages=messages,
+            max_tokens=max_tokens,
+            temperature=temperature,
+            response_model=response_model,
+            timeout=timeout,
+        )
+        response.data = instructor_response.model_dump(by_alias=True)
+    except Exception as e:
+        response.adala_error = True
+        response.adala_message = type(e).__name__
+        response.adala_details = traceback.format_exc()
+    finally:
+        return response.model_dump(by_alias=True)
```

Comment:
see above comment about separate data models for constrained and nonconstrained responses

Reply:
Resolved this with https://github.com/HumanSignal/Adala/pull/162#discussion_r1692436952